### ISSUE TITLE
Container doesn't start: bwrap: Can't mkdir /usr/lib32/gconv: Read-only file system

### ISSUE THREAD

**Issue by smcv** [ValveSoftware/steam-runtime#328]
From a comment by @DistantThunder on #321.

#### Reporter's system information

* Steam Runtime Version: **unknown**
* Distribution (e.g. Ubuntu 18.04): Arch Linux
* Link to your full system information (Help -> System Information) in a [Gist](https://gist.github.com/): **not provided**
* Have you checked for system updates?: **unspecified**
* Are you using the Steam Linux Runtime compatibility tool?: Yes

* System: Host: HOST Kernel: 5.9.13-3 x86_64 bits: 64 compiler: N/A Desktop: N/A Distro: Arch Linux
* CPU: Topology: 8-Core model: AMD Ryzen 7 3700X bits: 64 type: MT MCP arch: Zen L2 cache: 4096 KiB
* Graphics: Device-1: Advanced Micro Devices [AMD/ATI] Navi 10 [Radeon RX 5600 OEM/5600 XT / 5700/5700 XT]
* vendor: Sapphire Limited driver: amdgpu v: kernel bus ID: 0c:00.0
* Display: tty server: X.Org 1.20.10 driver: amdgpu resolution: 2560x1440~144Hz
* OpenGL: renderer: AMD Radeon RX 5700 XT (NAVI10 DRM 3.39.0 5.9.13-3-gc LLVM 11.0.0)
* v: 4.6 Mesa 21.0.0-devel (git-ac0d393eb1) direct render: Yes

#### Please describe your issue in as much detail as possible:

@DistantThunder writes:
>>>
I can't run anything with Proton 5.13.x:
```
pressure-vessel-wrap[3832930]: Replacing self with bwrap...
bwrap: Can't mkdir /usr/lib32/gconv: Read-only file system
ln: failed to create symbolic link '/tmp/SteamPVSockets.AVr7WY/SteamLinuxRuntime.6ec29bbe56106482/socket' -> '': No such file or directory
ERROR: ld.so: object '/home/_USER_/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/home/_USER_/.local/share/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
pressure-vessel-launch[3837936]: Can't connect to peer socket: Could not connect: No such file or directory
```
>>>

Log: https://github.com/ValveSoftware/steam-runtime/files/5685358/pv_c2077.problems.zip

#### Steps for reproducing this issue:

Not specified, but presumably: Run a Proton 5.13.x game.

**Comment by smcv** [comment]
@DistantThunder, we will need some more info on this. I've seen other reports with a similar error message, but none of them had enough information to pin down exactly what was wrong, which meant we didn't have the necessary information to figure out how to fix it.

I think I've also seen Fedora users reporting something similar with `/usr/lib64/gconv`, which could potentially be the same bug but in reverse (since the way Arch Linux does multilib is basically the reverse of how Fedora does it).

As a first thing, please try:

```
ls -ld /lib /lib32 /lib64 /usr/lib /usr/lib32 /usr/lib64 /usr/lib/gconv /usr/lib32/gconv /usr/lib64/gconv
```

and show us the output.

More generally, is there anything unusual about your system's setup or library search path that we should know about, or is it a fairly typical Arch installation?

**Comment by DistantThunder** [comment]
Hi @smcv, 

Please find below the requested details:

> $env LANG=C ls -ld /lib /lib32 /lib64 /usr/lib /usr/lib32 /usr/lib64 /usr/lib/gconv /usr/lib32/gconv /usr/lib64/gconv
ls: cannot access '/lib32': No such file or directory
lrwxrwxrwx   1 root root      7 Sep  3 00:30 /lib -> usr/lib
lrwxrwxrwx   1 root root      7 Sep  3 00:30 /lib64 -> usr/lib
drwxr-xr-x 357 root root 294912 Dec 13 15:37 /usr/lib
drwxr-xr-x   2 root root  12288 Oct 16 18:35 /usr/lib/gconv
drwxr-xr-x  46 root root  69632 Dec 13 15:37 /usr/lib32
drwxr-xr-x   2 root root  12288 Oct 16 18:36 /usr/lib32/gconv
lrwxrwxrwx   1 root root      3 Sep  3 00:30 /usr/lib64 -> lib
drwxr-xr-x   2 root root  12288 Oct 16 18:35 /usr/lib64/gconv


> Steam Runtime Version: ???
> Have you checked for system updates?: YES
> Link to your full system information (Help -> System Information) in a Gist: https://gist.github.com/DistantThunder/d8baff22460f4d9f753d33725a9e4cd8


> More generally, is there anything unusual about your system's setup or library search path that we should know about, or is it a fairly typical Arch installation?

My compat folder are under deduplicated ZFS file systems, it's the only oddity that I can think of.

I was testing ways to have Mangohud work under Proton 5.13.4 (it worked previously, before the game was updated) and I tried to remove the Soldier runtime and reinstall it: that's when I realized the games weren't working anymore with this Proton version.

**Comment by smcv** [comment]
From the symptoms you reported, I'd assumed there would be symbolic links involved.

Which version of SteamLinuxRuntime_soldier are you using? (Look in `VERSIONS.txt` - probably best if you just paste all of it.)

Also please try this:

```
cd .../SteamLinuxRuntime_soldier
ls -ld soldier/files/lib32 soldier/files/lib64 soldier/files/lib
ls -ld soldier/files/lib32/gconv soldier/files/lib64/gconv soldier/files/lib/gconv
```

> My compat folder are under deduplicated ZFS file systems, it's the only oddity that I can think of.

I don't *think* that should affect anything, although I'm not very familiar with ZFS.

**Comment by DistantThunder** [comment]
> #Name	Version		Runtime	Runtime_Version	Comment
SteamLinuxRuntime	v0.20201124.0-0-gd107701			# Entry point scripts, etc.
pressure-vessel	0.20201124.0+srt1	scout	0.20201125.0	# pressure-vessel-bin.tar.gz
soldier	0.20201124.0	soldier	0.20201124.0	# com.valvesoftware.SteamRuntime.Platform-amd64,i386-soldier-runtime.tar.gz

---
> SteamLinuxRuntime_soldier]$ cd ../SteamLinuxRuntime_soldier
ls -ld soldier/files/lib32 soldier/files/lib64 soldier/files/lib
ls -ld soldier/files/lib32/gconv soldier/files/lib64/gconv soldier/files/lib/gconv
drwxr-xr-x 16 darkeox darkeox 4096 Dec 13 20:08 soldier/files/lib
drwxr-xr-x  2 darkeox darkeox 4096 Dec 13 20:08 soldier/files/lib32
drwxr-xr-x  2 darkeox darkeox 4096 Dec 13 20:08 soldier/files/lib64
ls: cannot access 'soldier/files/lib32/gconv': No such file or directory
ls: cannot access 'soldier/files/lib64/gconv': No such file or directory
ls: cannot access 'soldier/files/lib/gconv': No such file or directory

**Comment by smcv** [comment]
OK, this is a bit weird. The archive `com.valvesoftware.SteamRuntime.Platform-amd64,i386-soldier-runtime.tar.gz` that gets unpacked to create the `soldier` directory definitely does contain `files/lib32/gconv`, `files/lib64/gconv` and `files/lib/gconv` directories (I downloaded a fresh copy of version 0.20201124.0 to verify, and they do exist).

@TTimo, when Steam updates a depot, I know it won't create empty directories that should have been in the depot (like git); but does it *delete* empty directories if they already *do* exist?

@DistantThunder, you can probably work around this by deleting the `soldier_0.20201124.0` directory. The next time you try to run a Proton 5.13 game, it will take longer than usual to start up while a new `soldier_0.20201124.0` directory is unpacked.

**Comment by TTimo** [comment]
@smcv Steam will only delete files that are present in the depot, it will ignore everything else. This looks like a duplicate of the other issues of partial/corrupted tarball expansion.

**Comment by DistantThunder** [comment]
That's what I was think as well, I checked disk space but I have and still have plenty where this is deployed.

https://framabin.org/p/?f96cda496074e1fd#8I7FaUXqIRkjbqpp3JpWoo65UXVYGENAgh0xQlqleqI=

It appears the files are indeed well in  the tarball. 

I tried @smcv fix and it worked:

>$ls -ld soldier/files/lib32/gconv soldier/files/lib64/gconv soldier/files/lib/gconv
drwxr-xr-x 2 _USER_ _USER_ 4096 Nov 24 16:15 soldier/files/lib/gconv
drwxr-xr-x 2 _USER_ _USER_ 4096 Nov 24 16:15 soldier/files/lib32/gconv
drwxr-xr-x 2 _USER_ _USER_ 4096 Nov 24 16:15 soldier/files/lib64/gconv

I can't tell you how the tarbal failed to uncompress however, it's the same tarball that I downloaded and everything went as smoothly as when I re-installed Soldier it seems, except this time around it worked...

**Comment by smcv** [comment]
> I can't tell you how the tarbal failed to uncompress however

No, we're not sure how this can happen either: even if it fails to uncompress once, it should recover automatically next time, because we don't rename the unpacked directory to its final name until *after* it should have already been fully unpacked. There must be some sort of logic error in the `run-in-steamrt` and `deploy-runtime` scripts, but I haven't been able to spot it.

**Comment by smcv** [comment]
> Mangohud still won't work but I guess this is a separate issue

Yes, #303.

**Comment by simonvanderveldt** [comment]
Seems like the same thing happened to me, no gconv directories. I was trying to install Natural Selection 2 using the current latest version of Proton (5.13.6).
One difference I noticed is that the paths that were mentioned in this issue are different for me, the `soldier_0.20210208.0` directory was in a `var` dir `steamapps/common/SteamLinuxRuntime_soldier/var/soldier_0.20210208.0`.

Not sure if relevant, but all my filesystems are ZFS, since in the issue description ZFS was also mentioned, maybe there's something happening there?

And finally deleting the `steamapps/common/SteamLinuxRuntime_soldier/var/soldier_0.20210208.0` directory and trying to start the game again seems to have extracted the files correct but still didn't give me a working game, it hangs as at the preparing popup.

**Comment by smcv** [comment]
> One difference I noticed is that the paths that were mentioned in this issue are different for me, the soldier_0.20210208.0 directory was in a var dir steamapps/common/SteamLinuxRuntime_soldier/var/soldier_0.20210208.0.

This just means you have a newer version. Older versions mixed up things managed by Steampipe and things that change at runtime in one directory; newer versions put all the variable/transient stuff in the `var/` subdirectory.

> Not sure if relevant, but all my filesystems are ZFS, since in the issue description ZFS was also mentioned, maybe there's something happening there?

If this ever happens again, please could you keep a copy of the broken directory instead of deleting it? Information about what is missing/wrong might help us to find out whether something is going wrong on ZFS specifically.

> And finally deleting the steamapps/common/SteamLinuxRuntime_soldier/var/soldier_0.20210208.0 directory and trying to start the game again seems to have extracted the files correct but still didn't give me a working game, it hangs as at the preparing popup.

Please see https://github.com/ValveSoftware/steam-runtime/blob/master/doc/reporting-steamlinuxruntime-bugs.md, and then report a separate issue with the requested information if what's logged doesn't match one of the known issues.

**Comment by Scrumplex** [comment]
I just backed up my existing `~/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/var/deploy-0.20210415.0/` and then deleted it which fixed the issue.

I generated the list of files using `find` and diffed it with the newly generated, working one:
https://gist.github.com/Scrumplex/04e55e8196d096df3dd33f0ba4695ef9

**Comment by smcv** [comment]
Versions >= 0.20210503.0 (currently only in the `client_beta` branch) have switched to a different way of distributing the runtime which should mean this failure mode can't happen.

**Comment by cagelight** [comment]
I've just recently started not being able to launch any proton games, getting this error. I switched to the beta branch of Steam (May 14 build) to see if it would fix but I still get failure to launch and `bwrap: Can't mkdir /usr/lib32/gconv: Read-only file system` in the console.

**Comment by Scrumplex** [comment]
cagelight: delete `~/.steam/steam/steamapps/common/SteamLinuxRuntime_soldier/var/deploy-<some_version_number>` and restart the game

**Comment by cagelight** [comment]
I did, and it fixed it, but I was posting that in response to the dev saying that it shouldn't be able to happen anymore. (unless it only happened cause of the cached data from the non-beta client)

**Comment by smcv** [comment]
@cagelight Please open a separate issue, with full logs if you can, so we can investigate.

**Comment by smcv** [comment]
@cagelight writes:
> I switched to the beta branch of Steam (May 14 build)

The beta branch of Steam does not change anything here. It's the beta branch of "Steam Linux Runtime - soldier" that would avoid this. Did you change the branch of "Steam Linux Runtime - soldier"? https://github.com/ValveSoftware/steam-runtime/blob/master/doc/reporting-steamlinuxruntime-bugs.md#using-a-beta-or-an-older-version

**Comment by cagelight** [comment]
Oh, my mistake, I read `client_beta` and assumed it was just the Steam Client Beta. Apologies for not finding that document.

**Comment by lectrode** [comment]
Ran into this same issue a few days ago (missing folder). Fixed it thanks to this thread.

In my case, this was likely the result of a script I ran to remove empty folders. If someone is deleting important files, that's on them, but I've always regarded deleting empty folders (especially if they are under the user's home directory) to be benign.

While certainly not required, it would be nice if this component was robust enough to handle creating missing folders (especially since newer proton is unusable without it). The runtime already uses bash scripts - I wouldn't think it'd be terribly difficult to add a directory check.

Or, maybe add a text file to those folders in the downloaded archive so they're not mistakenly deleted after being extracted.

**Comment by cagelight** [comment]
@lectrode Interesting that you mention a script to delete empty directories, I have one of those too and I got this issue after I ran it, so that must be the cause.

**Comment by smcv** [comment]
> I've always regarded deleting empty folders (especially if they are under the user's home directory) to be benign

Yes and no. It's more about how the relevant directory is managed and who manages it than about whether it's below $HOME.

If a directory is "yours", you can create and delete all the empty directories you want to, but if it's managed by something else (like Steam's CDN download system), please assume that the "something else" knows what it's doing, and don't arbitrarily delete directories that it has created - they are probably necessary (as in this case).

The on-disk size of an empty directory is very small, so you are not gaining much by deleting them, and you'll break programs that created them for a reason.

> it would be nice if this component was robust enough to handle creating missing folders

The current beta of Steam Linux Runtime - soldier happens to have unrelated changes to create directories programmatically rather than unpacking them from an archive, because we changed the way the runtime's files are distributed through the Steam CDN - so this should be fixed permanently in the beta.

This was done to consume less disk space, rather than specifically to solve this issue; and we were originally going to rely on the directories being created by the Steam CDN, but creating them programmatically from a list in a file turns out to be faster than reading the directory structure from disk if you are using a slow HDD.

> The runtime already uses bash scripts - I wouldn't think it'd be terribly difficult to add a directory check

Each incremental thing we add to a shell script is not difficult, but they quickly build up into something unmaintainable, so we're trying hard to avoid piling on checks for every possible way someone might have modified their system.

**Comment by kisak-valve** [comment]
Tracking note: Steam Linux Runtime (scout) mainline release has been bumped to 0.20210518.3 and Steam Linux Runtime - Soldier is at 0.20210415.0, so we're waiting on the latter to get updated past 0.20210503.0 to close this issue report.

**Comment by kisak-valve** [comment]
Steam Linux Runtime - Soldier mainline release has been bumped to 0.20210629.0. We're done here.
